cli/lockfile.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::path::PathBuf;
+
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_core::parking_lot::MutexGuard;
+use deno_core::ModuleSpecifier;
+use deno_runtime::deno_node::PackageJson;
+
+use crate::args::ConfigFile;
+use crate::cache;
+use crate::util::fs::atomic_write_file_with_retries;
+use crate::Flags;
+
+use crate::args::DenoSubcommand;
+use crate::args::InstallFlags;
+use crate::args::InstallKind;
+
+use deno_lockfile::Lockfile;
+
+#[derive(Debug)]
+pub struct CliLockfile {
+  lockfile: Mutex<Lockfile>,
+}
+
+pub struct Guard<'a, T> {
+  guard: MutexGuard<'a, T>,
+}
+
+impl<'a, T> std::ops::Deref for Guard<'a, T> {
+  type Target = T;
+
+  fn deref(&self) -> &Self::Target {
+    &self.guard
+  }
+}
+
+impl<'a, T> std::ops::DerefMut for Guard<'a, T> {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    &mut self.guard
+  }
+}
+
+impl CliLockfile {
+  pub fn new(lockfile: Lockfile) -> Self {
+    Self {
+      lockfile: Mutex::new(lockfile),
+    }
+  }
+
+  pub fn specifier(&self) -> Option<ModuleSpecifier> {
+    let path = &self.lockfile.lock().filename;
```

Comment:
Maybe we should store `filename` on self since it never changes? That also means we could skip the lock a bunch of times.

Reply:
lol I originally had that and then got rid of it because we don't actually use it many places. i'll put it back though, might as well avoid a few locks